src/script/engine.zig

```diff
@@ -358,6 +363,82 @@ pub const Engine = struct {
         try self.stack.pushByteArray(value);
     }
 
+    /// OP_NIP: Removes the second-to-top stack item
+    ///
+    /// # Returns
+    /// /// - "EngineError.StackUnderflow": if initial stack length < 2
+    fn opNip(self: *Engine) !void {
+        if (self.stack.len() < 2) {
+            return error.StackUnderflow;
+        }
+        const top_value = try self.stack.pop();
+        const second_to_top_value = try self.stack.pop();
+        try self.stack.pushByteArray(top_value);
```

Comment:
If elements on the stack are just pointers to allocated memory somewhere (i think it's the case in this implementation), you should just push back the pointer directly.
Here my understanding is that
`try self.stack.pushByteArray(top_value);`
will reallocate on the heap the exact same value you are freeing there
`defer self.allocator.free(top_value);`

You should reuse it

Reply:
If I remove the line:
`defer self.allocator.free(top_value);`

There is a leak in the memory.
My understanding is `pop` takes out the original data. `peek` returns a pointer

Reply:
As you can see [here](https://github.com/zig-bitcoin/btczee/blob/fa5329f4a74e326ead265ecad1baecde0989410b/src/script/stack.zig#L53) the function creates a copy and allocates that. the original remains as is and needs to be freed manually

Reply:
Yep, what I'm saying you is create another copy, sibling to `pushByteArray`, which will not copy the slice content into a newly allocated one, but rather store the already allocated one directly into the slice.

Instead of doing:
pop A
allocate B
copy A in B
push B
free A

You will be able to do:
pop A
push A

The stack content will be the same and you will have spared 1 free and 1 alloc